### PR TITLE
Adding padding preprocessing

### PR DESCRIPTION
**Pull request by gkroiz**
Added booleans `pad_data` for `prepare_alpaca.py` and `data_padded` for `finetune_adapter.py` that gives the user the option to add padding to all data in the preprocessing steps. If so, user has the option to remove padding when getting data during fine-tuning. 

The rational for this is that running with Pytorch XLA, the input data needs to keep the same dimensions to prevent recompilation. Open to suggestions on how to best add this.

### CODE REVIEW

**Review by carmocca — APPROVED** [review]
This looks good to me. Thoughts @awaelchli?

File: finetune_adapter.py [hunk 1]
```diff
@@ -32,6 +32,8 @@
 max_seq_length = 256  # see scripts/prepare_alpaca.py
 warmup_steps = epoch_size * 2 // micro_batch_size // devices  # 2 epochs
 
+data_padded = False
```
[the remark this answers is not in the document]
Comment (gkroiz) [reply finetune_adapter.py:35]: Thanks for the review @awaelchli. `block_size` is defined as `max_seq_length`. Since the `max_seq_length` is constant, doesn't it make sense to just preprocess the samples so they are all padded to the `max_seq_length` before training rather than padding them to `max_seq_length` each time we gather & prepare a batch?

[the remark this answers is not in the document]
Comment (awaelchli) [reply finetune_adapter.py:35]: This closely follows the nano-gpt style of data loading. We generally favor simplicity over highly optimized code. I believe the flexibility of being able to choose the max_seq_length and batch size in the script directly is very valuable and needed for users who want to get these scripts running on consumer hardware. We're not seeing any noticable data loading bottlenecks and therefore keep the current simple logic with memory mapping. 

I believe the padding to the constant shape should avoid the re-compilation, is that correct?

[the remark this answers is not in the document]
Comment (gkroiz) [reply finetune_adapter.py:35]: Yes, padding to a constant shape should avoid the re-compilation. I can revert the changes I made, and instead change the padding to a constant shape if using an xla device,

File: finetune_adapter.py [hunk 2]
```diff
@@ -174,11 +174,12 @@ def get_batch(fabric: L.Fabric, data: list):
     input_ids = [data[i]["input_ids"].type(torch.int64) for i in ix]
     labels = [data[i]["labels"].type(torch.int64) for i in ix]
 
-    max_len = max(len(s) for s in input_ids)
+    if data.device.type != "xla":
+        max_seq_length = max(len(s) for s in input_ids)
```
Comment (awaelchli) [inline finetune_adapter.py:178]:
```suggestion
    max_len = max(len(s) for s in input_ids) if data.device.type != "xla" else max_seq_length
```
I propose this instead of modifying the max_seq_length that is also global

**Review by awaelchli — APPROVED** [review]
Looks great. I only left one minor comment.

File: finetune_adapter.py [hunk 3]
```diff
@@ -174,11 +174,12 @@ def get_batch(fabric: L.Fabric, data: list):
     input_ids = [data[i]["input_ids"].type(torch.int64) for i in ix]
     labels = [data[i]["labels"].type(torch.int64) for i in ix]
 
-    max_len = max(len(s) for s in input_ids)
+    if data.device.type != "xla":
```
Comment (carmocca) [inline finetune_adapter.py:177]: please add a comment that the padding is done to avoid recompilation on xla

**Comment by gkroiz** [comment]
I added a commit to the `padding` branch, but it is not updating in this PR. Is this because this repo is no longer a fork of Lit-Llama?

**Comment by carmocca** [comment]
Possibly, we just made the fork change cc @lantiga 

Re-opening is totally fine

**Comment by gkroiz** [comment]
I'm planning on adding some more changes specific to TPUs for finetuning and since this change is only for xla devices, I will add it there. cc @carmocca
